### PR TITLE
Add Support for Parquet, File-Parallelism + Random-Read

### PR DESCRIPTION
**Pull request by JGoodlad**

### CODE REVIEW

File: MaxText/configs/base.yml [hunk 1]
```diff
@@ -371,4 +371,6 @@ dataset_directory: "/mnt/gcsfuse"
 gcs_metrics_bucket: "distributed-training-metrics"
 ### Parameters used for GCSFuse
 # The GCS bucket that holds the dataset.
-dataset_bucket: xai-hf-dataset-parquet
\ No newline at end of file
+dataset_bucket: xai-hf-dataset-parquet
+# The strategy used for reading the dataset.
```
Comment (bernardhan33) [inline MaxText/configs/base.yml:375]: nit: point to the location that lists the enums of these strategies.

File: MaxText/torch_datasets/parquet.py [hunk 2]
```diff
@@ -14,49 +14,113 @@
...
+      - Parquet metadata for all row groups is at the beginning of the file which
+        makes getting the metadata for row groups cheap.
+
+    Some implementation details:
+      - Row groups that are not a multiple of the batch size will return partial batches
+        on the last read of the row group.
+      - The reading of the row groups within a parquet file is random.
+    """
+    for single_parquest_file in assigned_parquet_files:
+      parquet_file = pq.ParquetFile(single_parquest_file, buffer_size = 0, pre_buffer = 0)
```
Comment (bernardhan33) [inline MaxText/torch_datasets/parquet.py:94]: Would be great to add a comment (link source, if available) on why these buffer_size , pre_buffer are 0.

File: MaxText/torch_datasets/parquet.py [hunk 3]
```diff
@@ -14,49 +14,113 @@
...
+      # Randomize the row groups
+      randomized_row_groups = self._random_iterator(range(parquet_file.num_row_groups))
+      for current_row_group in randomized_row_groups:
+        max_logging.log(f"row_group {current_row_group} has {parquet_file.metadata.row_group(current_row_group).num_rows} rows")
+        # TODO: Read batch size until all row groups have been read.
```
Comment (bernardhan33) [inline MaxText/torch_datasets/parquet.py:99]: Can we remove the TODO now? If this needs to be there, can we file a ticket to track it?

**Review by bernardhan33 — APPROVED** [review]
This is absolutely awesome work! Nice refactoring and code written for the enum, etc.

File: MaxText/configs/base.yml (hunk 1 above)
Comment (JGoodlad) [reply MaxText/configs/base.yml]: Done.

File: MaxText/torch_datasets/parquet.py (hunk 3 above)
Comment (JGoodlad) [reply MaxText/torch_datasets/parquet.py]: Sure! Removed.

File: MaxText/torch_datasets/parquet.py (hunk 2 above)
Comment (JGoodlad) [reply MaxText/torch_datasets/parquet.py:93]: I added the default parameters to make it clear that buffering would not be used (it is also not used by default but a question came up about buffering while I was working on this).

Added a comment to reflect this motivation.

File: MaxText/torch_datasets/parquet.py [hunk 4]
```diff
@@ -14,17 +14,86 @@
...
+    # batch_size is basically chunk size
+    self.allocated_parquet_files = allocated_parquet_files
+    self.columns = columns
+    self.batch_size = batch_size
+
+  def __iter__(self):
+    """File Parallel Random Read iterator.
+
+    Some facts:
+      - Parquet metadata for all row groups is at the beginning of the file which
```
[the remark this answers is not in the document]
Comment (JGoodlad) [reply MaxText/torch_datasets/parquet.py]: You are right. Fixed.

**Pull request merged**